src/protocols/ForeignToplevel.cpp

```diff
@@ -168,3 +168,14 @@ void CForeignToplevelProtocol::destroyHandle(CForeignToplevelHandle* handle) {
 bool CForeignToplevelProtocol::windowValidForForeign(PHLWINDOW pWindow) {
     return validMapped(pWindow) && !pWindow->isX11OverrideRedirect();
 }
+
+PHLWINDOW CForeignToplevelProtocol::windowFromHandleResource(wl_resource* res) {
```

Comment:
this sucks, as it's slow and unintuitive. See e.g. `CWLSurface::fromResource` for how to do it right.

Reply:
alright, changed it. I also changed it for the wlr toplevel which already had this method